suite/tests/client-interface/retinstru.c

```diff
@@ -0,0 +1,172 @@
+    print("return\n");
+    print("All done\n");
+    return 0;
+}
+
+
+#else /* asm code *************************************************************/
+#include "asm_defines.asm"
+START_FILE
+
+#ifdef X64
+# define FRAME_PADDING 8
+#else
+/* Don't need to align, but we do need to keep the "add esp, 0" to make a legal
+ * SEH64 epilog.
+ */
+# define FRAME_PADDING 0
+#endif
+
+DECL_EXTERN(callee)
+
+#define FUNCNAME test_ret
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        /* push callee-saved registers */
+        PUSH_SEH(REG_XBX)
+        PUSH_SEH(REG_XBP)
+        PUSH_SEH(REG_XSI)
+        PUSH_SEH(REG_XDI)
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+        CALLC0(GLOBAL_REF(callee))
+        mov      REG_XAX, PTRSZ [REG_XSP - ARG_SZ]
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        pop      REG_XDI
+        pop      REG_XSI
+        pop      REG_XBP
+        pop      REG_XBX
+        ret
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+#define FUNCNAME test_iret
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        /* push callee-saved registers */
+        PUSH_SEH(REG_XBX)
+        PUSH_SEH(REG_XBP)
+        PUSH_SEH(REG_XSI)
+        PUSH_SEH(REG_XDI)
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+#ifdef X64
+        push     HEX(2b) /* typical %ss value */
+        lea      rax, [rsp + 8]
+        push     rax
+        pushfq
+        push     HEX(33) /* typical %cs value */
+#else
+        pushfd
+        push     cs
+#endif
+        call skip_iret
+     next_instr_iret:
+#ifdef X64
+        mov      REG_XAX, PTRSZ [REG_XSP - 5*ARG_SZ]
+#else
+        mov      REG_XAX, PTRSZ [REG_XSP - 3*ARG_SZ]
+#endif
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        pop      REG_XDI
+        pop      REG_XSI
+        pop      REG_XBP
+        pop      REG_XBX
+        ret
+     skip_iret:
+#ifdef X64
+        iretq /* yeah the default is ridiculously iretd */
+#else
+        iretd
+#endif
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+#define FUNCNAME test_far_ret
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        /* push callee-saved registers */
+        PUSH_SEH(REG_XBX)
+        PUSH_SEH(REG_XBP)
+        PUSH_SEH(REG_XSI)
+        PUSH_SEH(REG_XDI)
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+#ifdef X64
+        push     HEX(33) /* typical %cs value */
+#else
+        push     cs
+#endif
+        call skip_far
+     next_instr_far:
+        mov      REG_XAX, PTRSZ [REG_XSP - 2*ARG_SZ]
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        pop      REG_XDI
+        pop      REG_XSI
+        pop      REG_XBP
+        pop      REG_XBX
+        ret
+     skip_far:
+#ifdef X64
+        RAW(48) /* yeah the default is ridiculously 4 bytes */
+#endif
+        RAW(cb) /* far ret */
+        END_FUNC(FUNCNAME)
+
+END_FILE
```

Comment:
This looks identical to retaddr.c.  Given that the test in the .dll.c is only counting and not checking the params, I would say let's not add a new identical app and just add the mbr code to retaddr.dll.c (where it will also function as a test of combining the mbr with the zero-retaddr instru).  (Or, add it to the count-ctis test which is what is already doing this kind of test for mbr instru, along with an iret in count-ctis.c.)